gloo/transport/tcp/listener.cc

```diff
@@ -0,0 +1,123 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <gloo/transport/tcp/listener.h>
+
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <iostream>
+
+#include <gloo/common/common.h>
+#include <gloo/common/logging.h>
+#include <gloo/transport/tcp/helpers.h>
+
+namespace gloo {
+namespace transport {
+namespace tcp {
+
+Listener::Listener(std::shared_ptr<Loop> loop, const attr& attr)
+    : loop_(std::move(loop)) {
+  listener_ = Socket::createForFamily(attr.ai_addr.ss_family);
+  listener_->reuseAddr(true);
+  listener_->bind(attr.ai_addr);
+  listener_->listen(kBacklog);
+  addr_ = listener_->sockName();
+
+  // Register with loop for readability events.
+  loop_->registerDescriptor(listener_->fd(), EPOLLIN, this);
+}
+
+Listener::~Listener() {
+  if (listener_) {
+    loop_->unregisterDescriptor(listener_->fd());
+  }
+}
+
+void Listener::handleEvents(int /* unused */) {
+  std::lock_guard<std::mutex> guard(mutex_);
```

Comment:
This is fine for now, but I'm wondering if the code is locking more than it should, as it calls into syscalls and handlers with the lock.

Reply:
It is probably possible to make this more fine grained, but the mutual exclusion here is between the I/O thread and a single user side driver thread. The user side thread will either wait for work to complete (and needs to acquire the mutex) or will post new work (and needs the mutex as well). In either case, it has to wait for a loop tick before new work can be handled, so it doesn't matter if the lock is acquired a bit later than strictly possible (due to coarse grained locking), because it would subsequently have to wait for the event loop thread again anyway.